File: wpimath/src/test/native/cpp/trajectory/proto/TrajectoryStateProtoTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include <gtest/gtest.h>

#include "frc/trajectory/Trajectory.h"

using namespace frc;

namespace {

using ProtoType = wpi::Protobuf<frc::Trajectory::State>;

const Trajectory::State kExpectedData = Trajectory::State{
    1.91_s, 4.4_mps, 17.4_mps_sq,
    Pose2d{Translation2d{1.74_m, 19.1_m}, Rotation2d{22.9_rad}},
    units::curvature_t{0.174}};
}  // namespace

TEST(TrajectoryStateProtoTest, Roundtrip) {
  google::protobuf::Arena arena;
  google::protobuf::Message* proto = ProtoType::New(&arena);
  ProtoType::Pack(proto, kExpectedData);

  Trajectory::State unpacked_data = ProtoType::Unpack(*proto);
  EXPECT_EQ(kExpectedData.t.value(), unpacked_data.t.value());
  EXPECT_EQ(kExpectedData.velocity.value(), unpacked_data.velocity.value());
  EXPECT_EQ(kExpectedData.acceleration.value(),
            unpacked_data.acceleration.value());
  EXPECT_EQ(kExpectedData.pose, unpacked_data.pose);
  EXPECT_EQ(kExpectedData.curvature.value(), unpacked_data.curvature.value());
}

File: wpimath/src/test/java/edu/wpi/first/math/trajectory/proto/TrajectoryStateProtoTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package org.wpilib.math.trajectory.proto;

import static org.junit.jupiter.api.Assertions.assertEquals;

import org.wpilib.math.geometry.Pose2d;
import org.wpilib.math.geometry.Rotation2d;
import org.wpilib.math.geometry.Translation2d;
import org.wpilib.math.proto.Trajectory.ProtobufTrajectoryState;
import org.wpilib.math.trajectory.Trajectory;
import org.junit.jupiter.api.Test;

class TrajectoryStateProtoTest {
  private static final Trajectory.State DATA =
      new Trajectory.State(
          1.91, 4.4, 17.4, new Pose2d(new Translation2d(1.74, 19.1), new Rotation2d(22.9)), 0.174);

  @Test
  void testRoundtrip() {
    ProtobufTrajectoryState proto = Trajectory.State.proto.createMessage();
    Trajectory.State.proto.pack(proto, DATA);

    Trajectory.State data = Trajectory.State.proto.unpack(proto);
    assertEquals(DATA.timeSeconds, data.timeSeconds);
    assertEquals(DATA.velocityMetersPerSecond, data.velocityMetersPerSecond);
    assertEquals(DATA.accelerationMetersPerSecondSq, data.accelerationMetersPerSecondSq);
    assertEquals(DATA.poseMeters, data.poseMeters);
    assertEquals(DATA.curvatureRadPerMeter, data.curvatureRadPerMeter);
  }
}
